services/eventservice/src/data_manager.rs

```diff
@@ -0,0 +1,301 @@
+        let org_id = event.organizer.to_be_bytes().clone();
+        let cat = event.category.as_bytes();
+        events.remove(id)?;
+        let organizers = self.db.open_tree(&self.organizers_name)?;
+        let categories = self.db.open_tree(&self.categories_name)?;
+        if organizers.contains_key(&org_id)? {
+            let ids = organizers.get(&org_id)?.unwrap();
+            let new_ids = EventManager::_remove_id(&ids[..], &id);
+            if new_ids.is_some() {
+                organizers.compare_and_swap(org_id, Some(ids), new_ids)??;
+            }
+        }
+        if categories.contains_key(&cat)? {
+            let ids = categories.get(&cat)?.unwrap();
+            let new_ids = EventManager::_remove_id(&ids[..], &id);
+            if new_ids.is_some() {
+                categories.compare_and_swap(cat, Some(ids), new_ids)??;
+            }
+        }
+        Ok(())
+    }
+
+    pub fn update_event(&self, new_event: &Event) -> Result<(), Box<dyn std::error::Error>> {
+        let id = new_event.id.to_be_bytes().clone();
+        let events = self.db.open_tree(&self.events_name)?;
+        events.insert(id, new_event.to_json().unwrap().as_bytes())?;
+        Ok(())
+    }
+
+    pub fn get_event(&self, event_id: &u64) -> Result<Event, Box<dyn std::error::Error>> {
+        let id = event_id.to_be_bytes();
+        let events = self.db.open_tree(&self.events_name)?;
+        let event = events.get(id)?;
+        if event.is_none() {
+            return Err(Box::new(std::io::Error::new(
+                std::io::ErrorKind::NotFound,
+                "Event not found",
+            )));
+        }
+        Event::from_json(&String::from_utf8(event.unwrap().to_vec())?)
+    }
+
+    fn _events_for_organizer(
+        &self,
+        organizer: u64,
+    ) -> Result<Vec<u64>, Box<dyn std::error::Error>> {
+        let organizers = self.db.open_tree(&self.organizers_name)?;
+        let org_ids = organizers.get(organizer.to_be_bytes())?;
+        match org_ids {
+            Some(ids) => Ok(EventManager::_ids_to_vec(&ids)),
+            None => Ok(Vec::new()),
+        }
+    }
+
+    fn _events_for_category(
+        &self,
+        category: &String,
+    ) -> Result<Vec<u64>, Box<dyn std::error::Error>> {
+        let categories = self.db.open_tree(&self.categories_name)?;
+        let cat_ids = categories.get(category.as_bytes())?;
+        match cat_ids {
+            Some(ids) => Ok(EventManager::_ids_to_vec(&ids)),
+            None => Ok(Vec::new()),
+        }
+    }
+
+    pub fn get_events(
+        &self,
+        sort_key: Option<&String>,
+        organizer: Option<u64>,
+        category: Option<&String>,
+    ) -> Result<Vec<Event>, Box<dyn std::error::Error>> {
+        let events = self.db.open_tree(&self.events_name)?;
+        let return_events: Result<Vec<Event>, Box<dyn std::error::Error>> =
+            match (organizer, category) {
```

Comment:
You can replace this mess of matching `Result` and `Option` with smith like this:

```rust
        if organizer.is_none() && category.is_none() {
            self._get_all_events()
        }
        let org_events: HashSet<u64> = match organizer {
            Some(org) => self._events_for_organizer(org)?.into_iter().collect(),
            None => HashSet::new(),
        };
        let cat_events: HashSet<u64> = match category {
            Some(cat) => self._events_for_category(cat)?.into_iter().collect(),
            None => HashSet::new(),
        };
        let mut result: Vec<Event> = (&org_events & &cat_events)
            .iter()
            .filter_map(|v| {
                if let Ok(Some(f)) = events.get(v.to_be_bytes()) {
                    return Event::from_json(&String::from_utf8(f.to_vec()).unwrap()).ok();
                }
                None
            })
            .collect();
        result.sort_by(|a, b| {
            EventManager::_compare_by_val(sort_key, a, b).unwrap_or(std::cmp::Ordering::Equal)
        });

        Ok(result)
```

Reply:
It would be nice, but intersection with empty set will always return empty set, that's why I need to cover all four cases separately

Reply:
Well, that's a general attitude that will make code more readable, imho. 

You can build an intermediate HashSet basing on values (a small match here will be much nicer than the previous one)

Or provide separate methods for each case.

Reply:
Btw, @OlegAndrus12, what do you think about this, you didn't like the method too.

Reply:
You didn't like it when it was built on ifs, so I've changed it. I can't find any other way to make the method shorter

Reply:
```rust
        if organizer.is_none() && category.is_none() {
            self._get_all_events()
        }
        let o: HashSet<u64> = match organizer {
            Some(org) => self._events_for_organizer(org)?.into_iter().collect(),
            None => HashSet::new(),
        };
        let c: HashSet<u64> = match category {
            Some(cat) => self._events_for_category(cat)?.into_iter().collect(),
            None => HashSet::new(),
        };
       let intersect_events = match (c, o) {
        (Some(c), Some(o)) => (&c & &o),
        (None, Some(o) => o,
        (Some(c), None) => c,
         (None, None) => HashSet::new()
       }
        let mut result: Vec<Event> = 
            .iter()
            .filter_map(|v| {
                if let Ok(Some(f)) = events.get(v.to_be_bytes()) {
                    return Event::from_json(&String::from_utf8(f.to_vec()).unwrap()).ok();
                }
                None
            })
            .collect();
        result.sort_by(|a, b| {
            EventManager::_compare_by_val(sort_key, a, b).unwrap_or(std::cmp::Ordering::Equal)
        });

        Ok(result)
```

So, we can't avoid this match against containers, but we can make it much smaller and simpler. So it will be easy to read and understand.

This will not make the method smaller, but easier to understand.

Reply:
Done